File: leet-code/lowestCommonAncestor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function (root, p, q) {
  if (root == null) {
    return null
  }
  while (root != null) {
    if (root.val > p.val && root.val > q.val) {
      root = root.left
    } else if (root.val < p.val && root.val < q.val) {
      root = root.right
    } else if (
      (root.val > q.val && root.val < p.val) ||
      (root.val > p.val && root.val < q.val)
    ) {
      return root
    } else if (root.val === p.val) {
      return p
    } else if (root.val === q.val) {
      return q
    }
  }
}

File: leet-code/lowestCommonAncestor.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	for root != nil {
		if root.Val > p.Val && root.Val > q.Val {
			root = root.Left
		} else if root.Val < p.Val && root.Val < q.Val {
			root = root.Right
		} else if (root.Val > q.Val && root.Val < p.Val) || (root.Val < q.Val && root.Val > p.Val) {
			return root
		} else if root.Val == q.Val {
			return q
		} else if root.Val == p.Val {
			return p
		}
	}
	return root
}
